Documentation: Authorization (#3470)
neo4j/graphql: docs/modules/ROOT/pages/guides/v4-migration/authorization.adoc

= Authentication and Authorization

> The largest breaking change in version 4.0.0 is the removal of the `@auth` directive, which will require a migration to the new `@authentication` and `@authorization` directives.

Reviewer: ```suggestion
The largest breaking change in version 4.0.0 is the removal of the `@auth` directive, which requires a migration to the new `@authentication` and `@authorization` directives.
```

== Instantiation

> Whilst the `@auth` directive required the installation of an additional plugin package, the functionality for the new directives is built directly into the library. You should uninstall the previous plugin:

Reviewer: ```suggestion
Whilst the `@auth` directive required the installation of an additional plugin package, the functionality for the new directives is built directly into the library. 
You should uninstall the previous plugin:
```

[source, bash, indent=0]
----
npm uninstall @neo4j/graphql-plugin-auth
----

Then, given an example of instantiation using a basic secret with the plugin:

[source, typescript, indent=0]
----
new Neo4jGraphQL({
    typeDefs,
    plugins: {
        auth: new Neo4jGraphQLAuthJWTPlugin({
            secret: "secret",
        }),
    }
})
----

You can delete the import of `Neo4jGraphQLAuthJWTPlugin` and change the instantiation to:

[source, typescript, indent=0]
----
new Neo4jGraphQL({
    typeDefs,
    features: {
        authorization: {
            key: "secret",
        }
    }
})
----

== Server

> Previously, you could pass in the entire request object and the library would fish the `Authorization` header out:

Reviewer: ```suggestion
Previously, you could pass in the entire request object and the library would find the `Authorization` header:
```

[source, typescript, indent=0]
----
const server = new ApolloServer({
    schema, // schema from Neo4jGraphQL.getSchema()
});

const { url } = await startStandaloneServer(server, {
    listen: { port: 4000 },
    context: async ({ req }) => ({ req }),
});
----

> With the new implementation, the library expects the `Authorization` header to be extracted in the format of a bearer token in the `token` field of the context:

Reviewer: Should "bearer". be capitalized?

Author: No, it's only capitalized when used inside a HTTP header as far as I'm aware!

[source, typescript, indent=0]
----
const server = new ApolloServer({
    schema, // schema from Neo4jGraphQL.getSchema()
});

const { url } = await startStandaloneServer(server, {
    listen: { port: 4000 },
    context: async ({ req }) => ({
        token: req.headers.Authorization,
    }),
});
----

> This is to acknowledge the fact that there are a variety of servers which don't have a `req` object (such as serverless functions, which use `event`). This is a uniform experience, no matter what server you use.

Reviewer: Could we replace uniform for default?

```suggestion
This is to acknowledge the fact that there are a variety of servers which don't have a `req` object (such as serverless functions, which use `event`). 
This is a uniform experience, no matter what server you use.
```

=== `rolesPath`

> `rolesPath` was used to configure a custom path for the "roles" claim in the JWT structure. This configuration has now been moved into the type definitions themselves. So given a previous instantiation:

Reviewer: ```suggestion
The `rolesPath` instantiation was used to configure a custom path for the "roles" claim in the JWT structure. 
This configuration has now been moved into the type definitions themselves. 
So given a previous instantiation:
```

[source, typescript, indent=0]
----
new Neo4jGraphQL({
    typeDefs,
    plugins: {
        auth: new Neo4jGraphQLAuthJWTPlugin({
            secret: "secret",
            rolesPath: "some.nested.path",
        }),
    }
})
----

> This will now need to instead be configured in the type definitions as:

Reviewer: ```suggestion
This now needs to instead be configured in the type definitions as:
```

[source, graphql, indent=0]
----
type JWT @jwt {
    roles: [String!]! @jwtClaim(path: "some.nested.path")
}
----

The type name itself can be anything, as long as it is decorated by `@jwt`.

> Whilst this is more boilerplate, the strongly typed nature of the definition means there is significantly more powerful filtering options in version 4.0.0.

Reviewer: Could we use a simpler word than boilerplate? I'm not sure how this is common among readers. Maybe "standardized"?

== Global authentication

> Global authentication was previously configured in the auth plugin consructor, for instance:

Reviewer: Should it be `@auth`?

Author: Not in this case! 🙂

[source, typescript, indent=0]
----
new Neo4jGraphQL({
    typeDefs,
    plugins: {
        auth: new Neo4jGraphQLAuthJWTPlugin({
            secret: "secret",
            globalAuthentication: true,
        }),
    }
})
----

To remain consistent with the use of directives for configuration, this is now achieved in type definitions by extending the schema:

[source, graphql, indent=0]
----
extend schema @authentication
----

== Rules

=== `allow`

Given an `allow` rule, which checks the `id` field of a `User` against the JWT subject _before_ any operation:

[source, graphql, indent=0]
----
type User @auth(rules: [{ allow: { id: "$jwt.sub" } }]) {
    id: ID!
}
----

> This will now become:

Reviewer: ```suggestion
This is now:
```

[source, graphql, indent=0]
----
type User @authorization(validate: [{ when: [BEFORE], where: { node: { id: "$jwt.sub" } } }]) {
    id: ID!
}
----

> Note that `allow` is no longer a discrete rule, but configured by a `when` argument which is an array accepting values `BEFORE` and `AFTER`.

Reviewer: ```suggestion
Note that `allow` is no longer a discrete rule, but configured by a `when` argument which is an array accepting the values `BEFORE` and `AFTER`.
```

> It is our view that users will quite rarely need to specify this argument as it defaults to both, and users will often want to validate a node property both before and after each operation.

Reviewer: We shouldn't use first person here, but this one is hard to be adapted since it's a disclosure of the team's decision. Maybe:

```suggestion
It is expected that users rarely need to specify this argument as it defaults to both.
Instead, users may prefer to validate a node property both before and after each operation.
```

Author: I made a slight tweak, with slightly different wording more similar to the original intention!

=== `bind`

Given an `bind` rule, which checks the `id` field of a `User` against the JWT subject _after_ any operation:

[source, graphql, indent=0]
----
type User @auth(rules: [{ bind: { id: "$jwt.sub" } }]) {
    id: ID!
}
----

This will now become:

[source, graphql, indent=0]
----
type User @authorization(validate: [{ when: [AFTER], where: { node: { id: "$jwt.sub" } } }]) {
    id: ID!
}
----

Note that `bind` is no longer a discrete rule, but configured by a `when` argument which is an array accepting values `BEFORE` and `AFTER`.

> It is our view that users will quite rarely need to specify this argument as it defaults to both, and users will often want to validate a node property both before and after each operation.

Reviewer: Same as before, in line 154

=== `isAuthenticated`

> WARNING: There isn't a direct replacement for the `isAuthenticated` argument, as you will read in this section. Please https://github.com/neo4j/graphql/issues/new/choose[raise a feature request] if this is an absolute blocker to migration.

Reviewer: ```suggestion
[WARNING]
====
There isn't a direct replacement for the `isAuthenticated` argument.
Please https://github.com/neo4j/graphql/issues/new/choose[raise a feature request] if this is blocking migration.
====
```

Given a previous type definition, which required authentication for any operation on the type `User`:

[source, graphql, indent=0]
----
type User @auth(rules: [{ isAuthenticated: true }]) {
    id: ID!
}
----

> There is not a rule under `@authorization` for this any more, but the closest is:

Reviewer: ```suggestion
There is not a rule under `@authorization` anymore, but the closest is:
```

[source, graphql, indent=0]
----
type User @authentication {
    id: ID!
}
----

The difference here being that for example, given the following query:

[source, graphql, indent=0]
----
{
    users(where: { id: "1" }) {
        id
    }
}
----

> * `@auth(rules: [{ isAuthenticated: true }])` would only throw an error if the `where: { id: "1" }` filter resulted in a match on a `User`.

Reviewer: ```suggestion
* `@auth(rules: [{ isAuthenticated: true }])` only throws an error if the `where: { id: "1" }` filter resulted in a match on a `User`.
```

> * `@authentication` will always throw an error if a user is not authenticated, and this will happen _before_ database execution to restrict database access to queries generated by authenticated users only.

Reviewer: ```suggestion
* `@authentication` always throws an error if a user is not authenticated.
This happens _before_ the database execution in order to restrict database access to queries generated by authenticated users only.
```

=== `roles`

[NOTE] 
====
For these examples, the following type is required in the type definitions:

[source, graphql, indent=0]
----
type JWT @jwt {
    roles: [String!]!
}
----
====

> Given the following type definition, which required a user to have the "admin" role to perform any operation on the type `User`:

Reviewer: Required or requires?

[source, graphql, indent=0]
----
type User @auth(rules: [{ roles: "admin" }]) {
    id: ID!
}
----

> This will now become:

Reviewer: ```suggestion
This is now:
```

[source, graphql, indent=0]
----
type User @authorization(validate: [{ where: { jwt: { roles_INCLUDES: "admin" } } }]) {
    id: ID!
}
----

> Note the following:

Author: I don't think it should be a note, so I have changed the phrasing slightly!

> * This is inside a `validate` rule, which will throw an error without the role as per the previous implementation. This can also be used inside a `filter` rule to just return zero results if a user does not have the required role.

Reviewer: ```suggestion
* The current example is inside a `validate` rule, which throws an error without the role as per the previous implementation. 
It can also be used inside a `filter` rule to just return zero results if a user does not have the required role.
```

Author: I haven't gone with the exact wording, but a bit of a tweak.

> * `roles` has become `roles_INCLUDES`, because the xref::filtering.adoc[full filtering capabilities of the library] can now be used within the `@authorization` directive.

Reviewer: ```suggestion
* `roles` has become `roles_INCLUDES`, because the xref::reference/filtering.adoc[full filtering capabilities of the library] can now be used within the `@authorization` directive.
```

> * `roles` is no longer a top-level rule field, but nested within `where` under `jwt` - any number of JWT claims can now be compared against, if configured within the type decorated with `@jwt`.

Reviewer: ```suggestion
* `roles` is no longer a top-level rule field, but nested within `where` under `jwt`.
Any number of JWT claims can now be compared against, if configured within the type decorated with `@jwt`.
```

=== `where`

> Perhaps the simplest migration, given an `@auth` rule which would have previously looked like:

Reviewer: We try not to assume what is simpler or easier for users

```suggestion
It replaces an `@auth` rule which would have previously looked like:
```
